### PR TITLE
[WIP] snippets refactoring

### PR DESCRIPTION
**Pull request by matej21**
Snippets are now completely independent on the nette/application. I will send related PR to the nette/application in few minutes/hours.

What do you think?

### CODE REVIEW

**Comment by dg** [comment]
@matej21 do you think it can be incorporated in 2.4?

**Comment by matej21** [comment]
@dg I have some ideas how to rewrite it a bit. But I'm not sure when I find the time :\ (maybe this weekend)

**Comment by matej21** [comment]
@dg reworked. One test is failing because of bad indent (not sure which indent is right, probably "actual" :) will fix later). nette/application#22 is also updated.

**Comment by dg** [comment]
Wow!

**Comment by matej21** [comment]
the indent is fixed. but there is a small BC break - previous behavior was a bit inconsistent and whitespaces on the last line were removed only for a snippet macro.

**Comment by f3l1x** [comment]
Good job @matej21! :+1: It looks more readable.

File: src/Latte/ISnippetBridge.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	 * @return bool
+	 */
+	public function isSnippetMode();
+
+
+	/**
+	 * @param string
+	 * @return bool
+	 */
+	public function isInvalid($name);
```
Comment (dg) [inline src/Latte/ISnippetBridge.php]: `needsRedraw`?

File: src/Latte/SnippetDriver.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
+<?php
```
Comment (dg) [inline src/Latte/SnippetDriver.php]: This should be IMHO in Latte\Runtime

File: src/Latte/Macros/BlockMacros.php [hunk 3]
```diff
@@ -201,17 +201,18 @@ public function macroBlock(MacroNode $node, PhpWriter $writer)
 				}
 				$parent->data->dynamic = TRUE;
 				$node->data->leave = TRUE;
-				$node->closingCode = "<?php \$this->global->dynSnippets[\$this->global->dynSnippetId] = ob_get_flush(); ?>";
+				$node->closingCode = "<?php \$this->global->snippetDriver->leave(); ?>";
+				$enterCode = "\$this->global->snippetDriver->enter(" . $writer->formatWord($name) . ", \\Latte\\SnippetDriver::TYPE_DYNAMIC);";
```
Comment (dg) [inline src/Latte/Macros/BlockMacros.php]: here are single quotes better

File: src/Latte/Macros/BlockMacros.php [hunk 4]
```diff
@@ -294,18 +295,14 @@ public function macroBlockEnd(MacroNode $node, PhpWriter $writer)
 				$node->content = $node->innerContent;
 			}
 
+			if (($node->name === 'snippet' || $node->name === 'snippetArea') && strpos($node->data->name, '$') === FALSE) {
+				$type = $node->name === 'snippet' ? 'TYPE_STATIC' : 'TYPE_AREA';
+				$node->content = '<?php $this->global->snippetDriver->enter('
```
Comment (dg) [inline src/Latte/Macros/BlockMacros.php]: what about `$type = $node->name === 'snippet' ? Latte\SnippetDriver::TYPE_STATIC : Latte\SnippetDriver::TYPE_AREA;`

File: src/Latte/SnippetDriver.php [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+class SnippetDriver
+{
+	use Strict;
+
+	const TYPE_STATIC = 'static',
+		TYPE_DYNAMIC = 'dynamic',
+		TYPE_AREA = 'area';
+
+	/** @var array */
+	protected $stack = [];
```
Comment (dg) [inline src/Latte/SnippetDriver.php]: better is to make everything private unless somebody need to make it protected

File: src/Latte/Template.php [hunk 6]
```diff
@@ -173,13 +173,18 @@ protected function initialize(& $params)
 				$this->renderBlock($this->params['_renderblock'], $this->params);
 				return TRUE;
 			}
+		} elseif (isset($this->global->snippetBridge) && !isset($this->global->snippetDriver)) {
```
Comment (dg) [inline src/Latte/Template.php]: Do you think that support for `$this->global->snippetBridge` is necessary?

File: src/Latte/Template.php [hunk 7]
```diff
@@ -173,13 +173,18 @@ protected function initialize(& $params)
...
+			$this->global->snippetDriver = new SnippetDriver($this->global->snippetBridge);
 		}
 
 		Runtime\Filters::$xhtml = (bool) preg_match('#xml|xhtml#', $this->contentType);
 		// old accumulators for back compatibility
 		$this->params['_l'] = $params['_l'] = new \stdClass;
 		$this->params['_g'] = $params['_g'] = $this->global;
 		$params['_b'] = (object) ['blocks' => & $this->blockQueue, 'types' => & $this->blockTypes];
+		if (isset($this->global->snippetDriver) && $this->global->snippetBridge->isSnippetMode()) {
```
Comment (dg) [inline src/Latte/Template.php]: `$this->global->snippetBridge` may not be set

File: tests/Latte/BlockMacros.snippet.alt.phpt [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+/**
+ * Test: general snippets test.
+ */
+
+use Nette\Utils\Html;
+use Nette\Bridges\ApplicationLatte\UIMacros;
```
Comment (dg) [inline tests/Latte/BlockMacros.snippet.alt.phpt]: UIMacros is needed?

**Comment by dg** [comment]
This look really cool, I'm :+1: for merge.

File: src/Latte/ISnippetBridge.php [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+	 */
+	public function getHtmlId($name);
+
+
+	/**
+	 * @param string snippet name
+	 * @param string html content
+	 * @return mixed
+	 */
+	public function addToPayload($name, $content);
```
Comment (dg) [inline src/Latte/ISnippetBridge.php]: `addSnippet`?

File: src/Latte/SnippetDriver.php [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+	/** @var int */
+	protected $renderingLevel = 0;
+
```
Comment (dg) [inline src/Latte/SnippetDriver.php]: $nestingLevel? or deep?

File: src/Latte/SnippetDriver.php [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+		if (($this->renderingLevel === 0 && $this->bridge->isInvalid($name))
+			|| ($type === self::TYPE_DYNAMIC && ($previous = end($this->stack)) && $previous[1] === TRUE)) {
+			ob_start(function () {});
+			$this->renderingLevel = $type === self::TYPE_AREA ? 0 : 1;
+			$obStarted = TRUE;
+		} elseif ($this->renderingLevel > 0) {
+			$this->renderingLevel++;
+		}
+		$this->stack[] = [$name, $obStarted];
+		if($name !== "") {
```
Comment (dg) [inline src/Latte/SnippetDriver.php]: Isn't it better to check for empty name inside bridge?

File: src/Latte/Macros/BlockMacros.php (hunk 4 above)
Comment (matej21) [inline src/Latte/Macros/BlockMacros.php]: value would be resolved in the compile time, not sure it is ok..

Comment (dg) [inline src/Latte/Macros/BlockMacros.php]: I know, but it is better to have readable source code that generated source code.

**Comment by matej21** [comment]
Updated 2b99d123044f
Is it ok now? if so, i'll squash it and let's merge! :)

**Comment by dg** [comment]
Great!

I am thinking … when SnippetDrivers is in Latte\Runtime, ISnippetBridge should be in Runtime too, or not? It is used only by Driver, so … probably should.

**Comment by matej21** [comment]
ok, updated

**Comment by dg** [comment]
Merged 3d1a58f75ac4bacbcd2154708e3c6e30596b9641 thanks!

**Comment by dg** [comment]
I don't know why, but this noticeably slows down web page with ublaboo/datagrid

**Comment by matej21** [comment]
In snippet mode or non-snippet mode? or both? will you investigate or should I?

**Comment by dg** [comment]
In non-snippet mode. I tried to use profiler, but with no success.

**Pull request closed without merging**